boot/freeldr/freeldr/lib/fs/ntfs.c

```diff
@@ -371,73 +373,182 @@ static ULONG NtfsReadAttribute(PNTFS_VOLUME_INFO Volume, PNTFS_ATTR_CONTEXT Cont
+        TRACE("Recursion = %u, AttrType = 0x%x\n", Recursion, AttrType);
+
+        if (ListContext)
+            NtfsReleaseAttributeContext(ListContext);
+        if (ListBuffer)
+            FrLdrTempFree(ListBuffer, TAG_NTFS_LIST);
+
+        ListContext = ListBuffer = NULL;
+
+        /* Limit the $ATTRIBUTE_LIST recursion or else infinity loop */
+        if (AttrType != Type && AttrType == NTFS_ATTR_TYPE_ATTRIBUTE_LIST && Recursion >= 1)
         {
-            PNTFS_ATTR_CONTEXT Context;
-            PNTFS_ATTR_CONTEXT ListContext;
-            PVOID ListBuffer;
-            ULONGLONG ListSize;
-            PNTFS_ATTR_RECORD ListAttrRecord;
-            PNTFS_ATTR_RECORD ListAttrRecordEnd;
+            PNTFS_ATTR_LIST_ATTR ListAttrRecord;
+            PNTFS_ATTR_LIST_ATTR ListAttrRecordEnd;
 
             ListContext = NtfsPrepareAttributeContext(AttrRecord);
 
             ListSize = NtfsGetAttributeSize(&ListContext->Record);
-            if(ListSize <= 0xFFFFFFFF)
+            if (ListSize <= 0xFFFFFFFF)
                 ListBuffer = FrLdrTempAlloc((ULONG)ListSize, TAG_NTFS_LIST);
             else
                 ListBuffer = NULL;
 
-            if(!ListBuffer)
+            if (!ListBuffer)
             {
                 TRACE("Failed to allocate memory: %x\n", (ULONG)ListSize);
-                continue;
+                goto skip;
             }
 
-            ListAttrRecord = (PNTFS_ATTR_RECORD)ListBuffer;
-            ListAttrRecordEnd = (PNTFS_ATTR_RECORD)((PCHAR)ListBuffer + ListSize);
+            ListAttrRecord = (PNTFS_ATTR_LIST_ATTR)ListBuffer;
+            ListAttrRecordEnd = (PNTFS_ATTR_LIST_ATTR)((PCHAR)ListBuffer + ListSize);
 
             if (NtfsReadAttribute(Volume, ListContext, 0, ListBuffer, (ULONG)ListSize) == ListSize)
             {
-                Context = NtfsFindAttributeHelper(Volume, ListAttrRecord, ListAttrRecordEnd,
-                                                  Type, Name, NameLength);
-
-                NtfsReleaseAttributeContext(ListContext);
-                FrLdrTempFree(ListBuffer, TAG_NTFS_LIST);
+                Context = NtfsFindAttributeHelperList(Volume, ListAttrRecord, ListAttrRecordEnd,
+                                                      Type, Name, NameLength,
+                                                      Recursion - 1);
 
                 if (Context != NULL)
-                    return Context;
+                    goto result;
             }
         }
 
-        if (AttrRecord->Type == Type)
+        if (AttrType == Type &&
+            AttrRecord->NameLength == NameLength &&
+            NtfsGetAttributeSize(AttrRecord) != 0 &&    /* HACK for ntfs3 driver on linux!!! because the ntfs3
+                                                         * likes to generate invalid attribute size when mft record
+                                                         * increases and i don't know how i will handle this */
```

Comment:
```suggestion
            /* HACK for ntfs3 driver on linux! Because the ntfs3 likes to generate invalid
             * attribute size when mft record increases and I don't know how I will handle this */
            NtfsGetAttributeSize(AttrRecord) != 0 &&
```